File: cmux-tui/bindings/typescript/src/protocol/topology.ts
import type { Id, SplitDirection } from "./common.js";

declare const uuidBrand: unique symbol;
/** A lowercase, hyphenated RFC 9562 UUID. */
export type UUID = string & { readonly [uuidBrand]: true };

export const TOPOLOGY_V8_CAPABILITIES = [
  "canonical-topology-snapshot-v1",
  "stable-entity-uuid-v1",
  "topology-resume-v1",
] as const;

export interface TopologyAuthority {
  daemon_instance_id: UUID;
  session_id: UUID;
}

export interface TopologyCursor extends TopologyAuthority { revision: number }

export type CanonicalLayout =
  | { type: "leaf"; pane: Id; pane_uuid: UUID }
  | { type: "split"; dir: SplitDirection; ratio: number; a: CanonicalLayout; b: CanonicalLayout };

export interface CanonicalTab {
  id: Id;
  uuid: UUID;
  kind: "pty" | "browser";
  name: string | null;
}

export interface CanonicalPane {
  id: Id;
  uuid: UUID;
  name: string | null;
  tabs: CanonicalTab[];
}

export interface CanonicalScreen {
  id: Id;
  uuid: UUID;
  name: string | null;
  layout: CanonicalLayout;
  panes: CanonicalPane[];
}

export interface CanonicalWorkspace {
  id: Id;
  uuid: UUID;
  name: string;
  screens: CanonicalScreen[];
}

export interface CanonicalTopology { workspaces: CanonicalWorkspace[] }

export interface TopologySnapshot extends TopologyCursor {
  topology: CanonicalTopology;
}

export interface TopologyTargets {
  workspaces: UUID[];
  screens: UUID[];
  panes: UUID[];
  surfaces: UUID[];
}

export type TopologyOperation =
  | "workspace-created" | "screen-created" | "pane-split" | "surface-attached"
  | "surface-closed" | "pane-closed" | "screen-closed" | "workspace-closed"
  | "workspace-renamed" | "screen-renamed" | "pane-renamed" | "surface-renamed"
  | "split-ratio-changed" | "panes-swapped" | "layout-applied" | "tab-moved"
  | "workspace-moved";

export interface TopologyDelta extends TopologyAuthority {
  event: "topology-delta";
  base_revision: number;
  revision: number;
  operation: TopologyOperation;
  targets: TopologyTargets;
  replacement: CanonicalTopology;
}

export type TopologyResnapshotReason =
  | "stale-daemon" | "stale-session" | "revision-ahead"
  | "history-gap" | "replay-too-large" | "slow-consumer";

export interface TopologyResnapshotRequiredResult extends TopologyAuthority {
  status: "resnapshot-required";
  current_revision?: number;
  reason: TopologyResnapshotReason;
}

export interface TopologyResnapshotRequiredEvent extends TopologyAuthority {
  event: "topology-resnapshot-required";
  current_revision?: number;
  reason: TopologyResnapshotReason;
}

export interface TopologySubscribedResult extends TopologyAuthority {
  status: "subscribed";
  from_revision: number;
  current_revision: number;
  replayed: number;
}

export type SubscribeTopologyResult = TopologySubscribedResult | TopologyResnapshotRequiredResult;
export type TopologyStreamEvent = TopologyDelta | TopologyResnapshotRequiredEvent;

export function topologyCursor(snapshot: TopologySnapshot): TopologyCursor {
  return {
    daemon_instance_id: snapshot.daemon_instance_id,
    session_id: snapshot.session_id,
    revision: snapshot.revision,
  };
}

export function parseTopologyCursor(value: unknown): TopologyCursor {
  const item = object(value, "topology cursor");
  return {
    daemon_instance_id: uuid(item.daemon_instance_id),
    session_id: uuid(item.session_id),
    revision: uint(item.revision),
  };
}

export function parseTopologySnapshot(value: unknown): TopologySnapshot {
  const item = object(value, "topology snapshot");
  return {
    daemon_instance_id: uuid(item.daemon_instance_id),
    session_id: uuid(item.session_id),
    revision: uint(item.revision),
    topology: canonicalTopology(item.topology),
  };
}

export function parseSubscribeTopologyResult(value: unknown): SubscribeTopologyResult {
  const item = object(value, "subscribe-topology result");
  const authority = {
    daemon_instance_id: uuid(item.daemon_instance_id),
    session_id: uuid(item.session_id),
  };
  if (item.status === "subscribed") {
    return {
      status: "subscribed",
      ...authority,
      from_revision: uint(item.from_revision),
      current_revision: uint(item.current_revision),
      replayed: uint(item.replayed),
    };
  }
  if (item.status === "resnapshot-required") {
    return {
      status: "resnapshot-required",
      ...authority,
      ...(item.current_revision === undefined ? {} : { current_revision: uint(item.current_revision) }),
      reason: resnapshotReason(item.reason),
    };
  }
  throw new Error(`invalid subscribe-topology status ${String(item.status)}`);
}

export function parseTopologyStreamEvent(value: unknown): TopologyStreamEvent {
  const item = object(value, "topology stream event");
  const authority = {
    daemon_instance_id: uuid(item.daemon_instance_id),
    session_id: uuid(item.session_id),
  };
  if (item.event === "topology-resnapshot-required") {
    return {
      event: "topology-resnapshot-required",
      ...authority,
      ...(item.current_revision === undefined ? {} : { current_revision: uint(item.current_revision) }),
      reason: resnapshotReason(item.reason),
    };
  }
  if (item.event !== "topology-delta") {
    throw new Error(`unexpected topology stream event ${String(item.event)}`);
  }
  return {
    event: "topology-delta",
    ...authority,
    base_revision: uint(item.base_revision),
    revision: uint(item.revision),
    operation: operation(item.operation),
    targets: targets(item.targets),
    replacement: canonicalTopology(item.replacement),
  };
}

export function validateTopologyDelta(
  cursor: TopologyCursor,
  delta: TopologyDelta,
): TopologyResnapshotRequiredEvent | null {
  let reason: TopologyResnapshotReason | null = null;
  if (delta.daemon_instance_id !== cursor.daemon_instance_id) reason = "stale-daemon";
  else if (delta.session_id !== cursor.session_id) reason = "stale-session";
  else if (delta.base_revision !== cursor.revision || delta.revision !== delta.base_revision + 1) {
    reason = "history-gap";
  }
  return reason === null ? null : {
    event: "topology-resnapshot-required",
    daemon_instance_id: delta.daemon_instance_id,
    session_id: delta.session_id,
    current_revision: delta.revision,
    reason,
  };
}

function canonicalTopology(value: unknown): CanonicalTopology {
  const item = object(value, "canonical topology");
  return { workspaces: array(item.workspaces, workspace) };
}

function workspace(value: unknown): CanonicalWorkspace {
  const item = object(value, "canonical workspace");
  return {
    id: uint(item.id), uuid: uuid(item.uuid), name: text(item.name),
    screens: array(item.screens, screen),
  };
}

function screen(value: unknown): CanonicalScreen {
  const item = object(value, "canonical screen");
  return {
    id: uint(item.id), uuid: uuid(item.uuid), name: nullableText(item.name),
    layout: layout(item.layout), panes: array(item.panes, pane),
  };
}

function layout(value: unknown): CanonicalLayout {
  const item = object(value, "canonical layout");
  if (item.type === "leaf") {
    return { type: "leaf", pane: uint(item.pane), pane_uuid: uuid(item.pane_uuid) };
  }
  if (item.type === "split" && (item.dir === "right" || item.dir === "down")) {
    return { type: "split", dir: item.dir, ratio: finite(item.ratio), a: layout(item.a), b: layout(item.b) };
  }
  throw new Error(`invalid canonical layout type ${String(item.type)}`);
}

function pane(value: unknown): CanonicalPane {
  const item = object(value, "canonical pane");
  return {
    id: uint(item.id), uuid: uuid(item.uuid), name: nullableText(item.name),
    tabs: array(item.tabs, tab),
  };
}

function tab(value: unknown): CanonicalTab {
  const item = object(value, "canonical tab");
  if (item.kind !== "pty" && item.kind !== "browser") throw new Error("invalid canonical tab kind");
  return { id: uint(item.id), uuid: uuid(item.uuid), kind: item.kind, name: nullableText(item.name) };
}

function targets(value: unknown): TopologyTargets {
  const item = object(value, "topology targets");
  return {
    workspaces: optionalArray(item.workspaces, uuid), screens: optionalArray(item.screens, uuid),
    panes: optionalArray(item.panes, uuid), surfaces: optionalArray(item.surfaces, uuid),
  };
}

function operation(value: unknown): TopologyOperation {
  const values: TopologyOperation[] = [
    "workspace-created", "screen-created", "pane-split", "surface-attached",
    "surface-closed", "pane-closed", "screen-closed", "workspace-closed",
    "workspace-renamed", "screen-renamed", "pane-renamed", "surface-renamed",
    "split-ratio-changed", "panes-swapped", "layout-applied", "tab-moved", "workspace-moved",
  ];
  if (!values.includes(value as TopologyOperation)) throw new Error(`invalid topology operation ${String(value)}`);
  return value as TopologyOperation;
}

function resnapshotReason(value: unknown): TopologyResnapshotReason {
  const values: TopologyResnapshotReason[] = [
    "stale-daemon", "stale-session", "revision-ahead", "history-gap", "replay-too-large", "slow-consumer",
  ];
  if (!values.includes(value as TopologyResnapshotReason)) throw new Error(`invalid resnapshot reason ${String(value)}`);
  return value as TopologyResnapshotReason;
}

function uuid(value: unknown): UUID {
  const textValue = text(value);
  if (!/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/.test(textValue)) {
    throw new Error(`invalid lowercase UUID ${textValue}`);
  }
  return textValue as UUID;
}

function object(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new Error(`${label} is not an object`);
  return value as Record<string, unknown>;
}
function array<T>(value: unknown, parse: (item: unknown) => T): T[] {
  if (!Array.isArray(value)) throw new Error("expected array");
  return value.map(parse);
}
function optionalArray<T>(value: unknown, parse: (item: unknown) => T): T[] {
  return value === undefined ? [] : array(value, parse);
}
function uint(value: unknown): number {
  if (!Number.isSafeInteger(value) || (value as number) < 0) throw new Error(`expected non-negative integer, got ${String(value)}`);
  return value as number;
}
function finite(value: unknown): number {
  if (typeof value !== "number" || !Number.isFinite(value)) throw new Error("expected finite number");
  return value;
}
function text(value: unknown): string {
  if (typeof value !== "string") throw new Error("expected string");
  return value;
}
function nullableText(value: unknown): string | null {
  return value === null ? null : text(value);
}

File: cmux-tui/bindings/python/cmux/topology.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any, Dict, List, Optional
from uuid import UUID


CANONICAL_TOPOLOGY_SNAPSHOT_CAPABILITY = "canonical-topology-snapshot-v1"
STABLE_ENTITY_UUID_CAPABILITY = "stable-entity-uuid-v1"
TOPOLOGY_RESUME_CAPABILITY = "topology-resume-v1"
TOPOLOGY_V8_CAPABILITIES = (
    CANONICAL_TOPOLOGY_SNAPSHOT_CAPABILITY,
    STABLE_ENTITY_UUID_CAPABILITY,
    TOPOLOGY_RESUME_CAPABILITY,
)


@dataclass(frozen=True)
class TopologyAuthority:
    daemon_instance_id: UUID
    session_id: UUID


@dataclass(frozen=True)
class TopologyCursor:
    daemon_instance_id: UUID
    session_id: UUID
    revision: int

    @property
    def authority(self) -> TopologyAuthority:
        return TopologyAuthority(self.daemon_instance_id, self.session_id)


@dataclass(frozen=True)
class CanonicalLayout:
    type: str
    pane: Optional[int] = None
    pane_uuid: Optional[UUID] = None
    dir: Optional[str] = None
    ratio: Optional[float] = None
    a: Optional["CanonicalLayout"] = None
    b: Optional["CanonicalLayout"] = None


@dataclass(frozen=True)
class CanonicalTab:
    id: int
    uuid: UUID
    kind: str
    name: Optional[str]


@dataclass(frozen=True)
class CanonicalPane:
    id: int
    uuid: UUID
    name: Optional[str]
    tabs: List[CanonicalTab]


@dataclass(frozen=True)
class CanonicalScreen:
    id: int
    uuid: UUID
    name: Optional[str]
    layout: CanonicalLayout
    panes: List[CanonicalPane]


@dataclass(frozen=True)
class CanonicalWorkspace:
    id: int
    uuid: UUID
    name: str
    screens: List[CanonicalScreen]


@dataclass(frozen=True)
class CanonicalTopology:
    workspaces: List[CanonicalWorkspace]


@dataclass(frozen=True)
class TopologySnapshot:
    daemon_instance_id: UUID
    session_id: UUID
    revision: int
    topology: CanonicalTopology

    @property
    def cursor(self) -> TopologyCursor:
        return TopologyCursor(self.daemon_instance_id, self.session_id, self.revision)


@dataclass(frozen=True)
class TopologyTargets:
    workspaces: List[UUID]
    screens: List[UUID]
    panes: List[UUID]
    surfaces: List[UUID]


class TopologyOperation(str, Enum):
    WORKSPACE_CREATED = "workspace-created"
    SCREEN_CREATED = "screen-created"
    PANE_SPLIT = "pane-split"
    SURFACE_ATTACHED = "surface-attached"
    SURFACE_CLOSED = "surface-closed"
    PANE_CLOSED = "pane-closed"
    SCREEN_CLOSED = "screen-closed"
    WORKSPACE_CLOSED = "workspace-closed"
    WORKSPACE_RENAMED = "workspace-renamed"
    SCREEN_RENAMED = "screen-renamed"
    PANE_RENAMED = "pane-renamed"
    SURFACE_RENAMED = "surface-renamed"
    SPLIT_RATIO_CHANGED = "split-ratio-changed"
    PANES_SWAPPED = "panes-swapped"
    LAYOUT_APPLIED = "layout-applied"
    TAB_MOVED = "tab-moved"
    WORKSPACE_MOVED = "workspace-moved"


@dataclass(frozen=True)
class TopologyDelta:
    daemon_instance_id: UUID
    session_id: UUID
    base_revision: int
    revision: int
    operation: TopologyOperation
    targets: TopologyTargets
    replacement: CanonicalTopology


class TopologyResnapshotReason(str, Enum):
    STALE_DAEMON = "stale-daemon"
    STALE_SESSION = "stale-session"
    REVISION_AHEAD = "revision-ahead"
    HISTORY_GAP = "history-gap"
    REPLAY_TOO_LARGE = "replay-too-large"
    SLOW_CONSUMER = "slow-consumer"


@dataclass(frozen=True)
class TopologyResnapshotRequired:
    daemon_instance_id: UUID
    session_id: UUID
    current_revision: Optional[int]
    reason: TopologyResnapshotReason


@dataclass(frozen=True)
class TopologySubscribed:
    daemon_instance_id: UUID
    session_id: UUID
    from_revision: int
    current_revision: int
    replayed: int


def parse_topology_snapshot(value: Dict[str, Any]) -> TopologySnapshot:
    return TopologySnapshot(
        daemon_instance_id=_uuid(value["daemon_instance_id"]),
        session_id=_uuid(value["session_id"]),
        revision=_revision(value["revision"]),
        topology=parse_canonical_topology(value["topology"]),
    )


def parse_topology_subscribed(value: Dict[str, Any]) -> TopologySubscribed:
    return TopologySubscribed(
        daemon_instance_id=_uuid(value["daemon_instance_id"]),
        session_id=_uuid(value["session_id"]),
        from_revision=_revision(value["from_revision"]),
        current_revision=_revision(value["current_revision"]),
        replayed=_revision(value["replayed"]),
    )


def parse_topology_delta(value: Dict[str, Any]) -> TopologyDelta:
    targets = value.get("targets", {})
    return TopologyDelta(
        daemon_instance_id=_uuid(value["daemon_instance_id"]),
        session_id=_uuid(value["session_id"]),
        base_revision=_revision(value["base_revision"]),
        revision=_revision(value["revision"]),
        operation=TopologyOperation(value["operation"]),
        targets=TopologyTargets(
            workspaces=_uuid_list(targets.get("workspaces", [])),
            screens=_uuid_list(targets.get("screens", [])),
            panes=_uuid_list(targets.get("panes", [])),
            surfaces=_uuid_list(targets.get("surfaces", [])),
        ),
        replacement=parse_canonical_topology(value["replacement"]),
    )


def parse_resnapshot_required(value: Dict[str, Any]) -> TopologyResnapshotRequired:
    current = value.get("current_revision")
    return TopologyResnapshotRequired(
        daemon_instance_id=_uuid(value["daemon_instance_id"]),
        session_id=_uuid(value["session_id"]),
        current_revision=None if current is None else _revision(current),
        reason=TopologyResnapshotReason(value["reason"]),
    )


def validate_topology_delta(
    cursor: TopologyCursor,
    delta: TopologyDelta,
) -> Optional[TopologyResnapshotRequired]:
    if delta.daemon_instance_id != cursor.daemon_instance_id:
        reason = TopologyResnapshotReason.STALE_DAEMON
    elif delta.session_id != cursor.session_id:
        reason = TopologyResnapshotReason.STALE_SESSION
    elif delta.base_revision != cursor.revision or delta.revision != delta.base_revision + 1:
        reason = TopologyResnapshotReason.HISTORY_GAP
    else:
        return None
    return TopologyResnapshotRequired(
        daemon_instance_id=delta.daemon_instance_id,
        session_id=delta.session_id,
        current_revision=delta.revision,
        reason=reason,
    )


def parse_canonical_topology(value: Dict[str, Any]) -> CanonicalTopology:
    return CanonicalTopology(
        workspaces=[_parse_workspace(item) for item in value.get("workspaces", [])]
    )


def _parse_workspace(value: Dict[str, Any]) -> CanonicalWorkspace:
    return CanonicalWorkspace(
        id=_revision(value["id"]),
        uuid=_uuid(value["uuid"]),
        name=str(value["name"]),
        screens=[_parse_screen(item) for item in value.get("screens", [])],
    )


def _parse_screen(value: Dict[str, Any]) -> CanonicalScreen:
    return CanonicalScreen(
        id=_revision(value["id"]),
        uuid=_uuid(value["uuid"]),
        name=value.get("name"),
        layout=_parse_layout(value["layout"]),
        panes=[_parse_pane(item) for item in value.get("panes", [])],
    )


def _parse_layout(value: Dict[str, Any]) -> CanonicalLayout:
    if value.get("type") == "leaf":
        return CanonicalLayout(
            type="leaf",
            pane=_revision(value["pane"]),
            pane_uuid=_uuid(value["pane_uuid"]),
        )
    if value.get("type") == "split":
        return CanonicalLayout(
            type="split",
            dir=str(value["dir"]),
            ratio=float(value["ratio"]),
            a=_parse_layout(value["a"]),
            b=_parse_layout(value["b"]),
        )
    raise ValueError(f"invalid canonical layout type {value.get('type')!r}")


def _parse_pane(value: Dict[str, Any]) -> CanonicalPane:
    return CanonicalPane(
        id=_revision(value["id"]),
        uuid=_uuid(value["uuid"]),
        name=value.get("name"),
        tabs=[_parse_tab(item) for item in value.get("tabs", [])],
    )


def _parse_tab(value: Dict[str, Any]) -> CanonicalTab:
    kind = str(value["kind"])
    if kind not in ("pty", "browser"):
        raise ValueError(f"invalid canonical tab kind {kind!r}")
    return CanonicalTab(
        id=_revision(value["id"]),
        uuid=_uuid(value["uuid"]),
        kind=kind,
        name=value.get("name"),
    )


def _uuid(value: Any) -> UUID:
    parsed = UUID(str(value))
    if str(parsed) != value:
        raise ValueError(f"UUID must use lowercase hyphenated form: {value!r}")
    return parsed


def _uuid_list(values: List[Any]) -> List[UUID]:
    return [_uuid(value) for value in values]


def _revision(value: Any) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or value < 0:
        raise ValueError(f"expected non-negative integer, got {value!r}")
    return value
